sound/soc/sof/intel/hda.c

```diff
@@ -251,15 +261,56 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	}
 
 	/* codec detection */
-	if (!bus->codec_mask)
+	if (!bus->codec_mask) {
 		dev_info(bus->dev, "no hda codecs found!\n");
-	else
-		dev_info(bus->dev, "hda codecs found, mask %lx!\n", bus->codec_mask);
+	} else {
+		dev_info(bus->dev, "hda codecs found, mask %lx\n",
+			 bus->codec_mask);
+
+		for (i = 0; i < HDA_MAX_CODECS; i++) {
+			if (bus->codec_mask & (1 << i))
+				codec_num++;
+		}
+
+		/*
+		 * If no machine driver is found, then:
+		 *
+		 * hda machine driver is used if :
+		 * 1. there are HDMI codec and other HDA codecs
```

Comment:
there is one HDMI codec and one external HDAudio codec.

Reply:
yes, I will refine it